src/hivpy/prep.py

```diff
@@ -746,25 +786,84 @@ def restart_prep(self, pop: Population, time_step):
                 # set start dates
                 self.set_all_prep_start_dates(pop, restarting_prep_pop)
                 # set continuous use
-                pop.set_present_variable(col.CONT_ON_PREP, time_step, restarting_prep_pop)
+                prep_cont = pop.get_variable(col.CONT_ON_PREP, restarting_prep_pop) + time_step
+                prep_intent_cont = pop.get_variable(col.CONT_INTENT_ON_PREP, restarting_prep_pop) + time_step
+                pop.set_present_variable(col.CONT_ON_PREP, prep_cont, restarting_prep_pop)
+                pop.set_present_variable(col.CONT_INTENT_ON_PREP, prep_intent_cont, restarting_prep_pop)
                 pop.set_present_variable(col.CONT_ACTIVE_ON_PREP, time_step, restarting_prep_pop)
                 # increment cumulative use
                 self.set_all_prep_cumulative(pop, restarting_prep_pop, time_step)
+                pop.set_present_variable(col.LAST_PREP_USE_DATE, pop.date, restarting_prep_pop)
                 # unset stop date
                 pop.set_present_variable(col.LAST_PREP_STOP_DATE, None, restarting_prep_pop)
+                # unpause prep
+                pop.set_present_variable(col.PREP_PAUSED, False, restarting_prep_pop)
 
-    def calc_restarting_prep(self, favoured_prep, size):
+    def calc_restarting_prep(self, favoured_prep, prep_paused, size):
         """
         Returns PrEP types for people restarting PrEP.
         Individual preferences and availability are taken into account.
         """
         # outcomes
         r = rng.uniform(size=size)
-        restarting = r < self.prob_prep_restart
+        # those who stopped due to ineligibility are guaranteed to restart
+        if prep_paused:
+            restarting = r < 1
+        else:
+            restarting = r < self.prob_prep_restart
         prep = [favoured_prep if r else None for r in restarting]
 
         return prep
 
+    def stop_prep(self, pop: Population, time_step):
+        """
+        Update PrEP usage for people stopping PrEP due to lack of eligibility
+        and people who have paused PrEP usage.
+        """
+        # people who are using prep and are now ineligible
+        ineligible = pop.get_sub_pop(AND(COND(col.PREP_ELIGIBLE, op.eq, False),
+                                         COND(col.EVER_PREP, op.eq, True),
+                                         COND(col.LAST_PREP_STOP_DATE, op.eq, None)))
+
+        if len(ineligible) > 0:
+            # reset active continuous use
+            pop.set_present_variable(col.CONT_ACTIVE_ON_PREP, timedelta(months=0), ineligible)
+            # set stop date
+            pop.set_present_variable(col.LAST_PREP_STOP_DATE, pop.date, ineligible)
+            # pause prep
```

Comment:
This will get updated for people who are paused as well since they are not eligible but have ever prep + no last prep stop date